src/test/java/com/znsio/teswiz/runner/AppPathTest.java

```diff
@@ -0,0 +1,105 @@
+        assertEquals(expectedAppPath, actualAppPath);
+        assertTrue(Files.exists(Paths.get(actualAppPath)));
+    }
+
+    @Test
+    void validateAppPathAsCorrectUrlOnlyDownloadFileWhenRepoExist() {
+        deleteFile(expectedAppPath);
+        createDirectory(expectedDirectoryPath);
+        String actualAppPath = DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        assertEquals(expectedAppPath, actualAppPath);
+        assertTrue(Files.exists(Paths.get(actualAppPath)));
+    }
+
+    @Test
+    void validateAppPathAsIncorrectUrlGivesError() {
+        deleteDirectorySubDirectoryAndFiles(tempFolderPath);
+        assertThrows(RuntimeException.class, () -> DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsIncorrectUrl));
+    }
+
+    @Test
+    void validateAppPathAsCorrectFilePathLoadsFileWhenAlreadyExist() {
+        DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        String actualAppPath = DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsFilePath);
+        assertEquals(expectedAppPath, actualAppPath);
+        assertTrue(Files.exists(Paths.get(actualAppPath)));
+    }
+
+    @Test
+    void validateAppPathAsIncorrectFilePathGivesError() {
+        DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        assertThrows(RuntimeException.class, () -> DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsIncorrectFilePath));
+    }
+
+    @Test
+    void validateAppPathAsFilePathButFileNotPresentGivesError() {
+        deleteFile(expectedAppPath);
+        assertThrows(RuntimeException.class, () -> DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsIncorrectFilePath));
+    }
+
+    private void deleteFile(String filePath) {
+        if (Files.exists(Path.of(filePath))) {
+            try {
+                Files.delete(Path.of(filePath));
+            } catch (IOException e) {
+                System.err.println("Failed to delete file for unit test: " + e.getMessage());
+            }
+        }
+    }
+
+    private void deleteDirectorySubDirectoryAndFiles(String directoryPath) {
+        if (Files.exists(Paths.get(directoryPath))) {
+            try {
+                Files.walk(Paths.get(directoryPath))
+                        .sorted(java.util.Comparator.reverseOrder())
+                        .map(Path::toFile)
+                        .forEach(java.io.File::delete);
+            } catch (IOException e) {
+                System.err.println("Failed to delete folder for unit test: " + e.getMessage());
+            }
+        }
+    }
+
+    private void createDirectory(String directoryPath) {
+        if (!Files.exists(Paths.get(directoryPath))) {
+            try {
+                Files.createDirectories(Paths.get(directoryPath));
+            } catch (IOException e) {
+                System.err.println("Failed to create directory: " + expectedDirectoryPath + " for unit test, error occurred" + e);
+            }
+        }
+    }
+}
```

Comment:
What if there is a file path given in the capabilities - which is valid, or invalid? Do we have tests for that?

Reply:
givenCorrectFilePath_WhenRepoAndFileAlreadyExist_ThenFileLoadsSuccessfully()
givenIncorrectFilePath_WhenRepoAndFileExist_ThenFileNotFoundExceptionOccursWhileTryingToOpenFile()
givenCorrectFilePath_WhenRepoExistButFileDoNotExist_ThenFileNotFoundExceptionOccursWhileTryingToOpenFile()
all these are for file path, other are for url